File: 01 - Classical/05 - Dynamic programming/07 - Max amount of gold/knapsack.cpp
#include <iostream>
#include <vector>



int optimal_weight(int W, const std::vector<int> &w) {

  // Given n gold bars, find the maximum weight of gold that fits in into bag ofcapacity W
  
  size_t n=w.size();

  std::vector <std::vector<int>> K(n+1,std::vector<int> (W+1));

  for (int i=0; i<=n; ++i){

    for (int weight=0; weight<=W; ++weight){

      if (i==0 || weight==0){

        K[i][weight]=0;

      } else if (w[i-1]<=weight){

        K[i][weight]=std::max(w[i-1]+K[i-1][weight-w[i-1]],K[i-1][weight]);

      } else {

        K[i][weight]=K[i-1][weight];

      }

    }

  }

  return K[n][W];

}

int main() {
  int n, W;
  std::cin >> W >> n;
  std::vector<int> w(n);
  for (int i = 0; i < n; i++) {
    std::cin >> w[i];
  }
  std::cout << optimal_weight(W, w) << '\n';
}

File: 01 - Classical/05 - Dynamic programming/07 - Max amount of gold/knapsack.py
# Uses python3
import numpy as np

def optimal_weight(W, w):
    
    '''
    
    Given n gold bars, find the maximum weight of gold that fits in into bag ofcapacity W
    
    '''
    
    global K
    
    K=np.zeros([n+1,W+1])
    
    
        
    for weight in range(W+1):
        
        for i in range(n+1):
            
            if (i==0 or weight==0):
                
                K[i][w]=0
            
            elif (w[i-1]<=weight):
                
                K[i][weight]=max(w[i-1]+K[i-1][weight-w[i-1]],K[i-1][weight])
            
            else:
                K[i][weight]=K[i-1][weight]
    
    return K[n][W]

if __name__ == '__main__':
    
    W=10
    n=3
    w=[1,4,8]
    
    print(optimal_weight(W, w))
    
